index.js

```diff
@@ -44,32 +44,43 @@ const TypedDataUtils = {
     const encodedTypes = ['bytes32']
     const encodedValues = [this.hashType(primaryType, types)]
 
-    for (const field of types[primaryType]) {
-      let value = data[field.name]
-      if (value !== undefined) {
-        if (field.type === 'bytes') {
-          encodedTypes.push('bytes32')
-          value = ethUtil.sha3(value)
-          encodedValues.push(value)
-        } else if (field.type === 'string') {
-          encodedTypes.push('bytes32')
-          // convert string to buffer - prevents ethUtil from interpreting strings like '0xabcd' as hex
-          if (typeof value === 'string') {
-            value = Buffer.from(value, 'utf8')
-          }
-          value = ethUtil.sha3(value)
-          encodedValues.push(value)
-        } else if (types[field.type] !== undefined) {
-          encodedTypes.push('bytes32')
-          value = ethUtil.sha3(this.encodeData(field.type, value, types))
-          encodedValues.push(value)
-        } else if (field.type.lastIndexOf(']') === field.type.length - 1) {
-          throw new Error('Arrays currently unimplemented in encodeData')
-        } else {
-          encodedTypes.push(field.type)
-          encodedValues.push(value)
+    const encodeField = (name, type, value) => {
+      if(value === undefined)
+        throw new Error(`missing value for field ${name} of type ${type}`)
```

Comment:
I've decided to turn missing values into an error. I'm not sure if it is proper to do so, or if values may be optional and skipped (how the implementation was before). If a decision is made on this, maybe a test case should be added to reflect that decision?

Reply:
This is probably worth asking on the EIP's discussion, to gauge for strong opinions.

Reply:
Okay I've done so: https://ethereum-magicians.org/t/eip-712-eth-signtypeddata-as-a-standard-for-machine-verifiable-and-human-readable-typed-data-signing/397/19?u=cag